src/block/Beacon.php

```diff
@@ -23,11 +23,209 @@
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use function array_merge;
+
 final class Beacon extends Transparent{
 
+	public const MIN_LEVEL_BEACON = 1;
+	public const MAX_LEVEL_BEACON = 4;
+
+	public const ALLOWED_ITEM_IDS = [
+		ItemTypeIds::IRON_INGOT => true,
+		ItemTypeIds::GOLD_INGOT => true,
+		ItemTypeIds::DIAMOND => true,
+		ItemTypeIds::EMERALD => true,
+		ItemTypeIds::NETHERITE_INGOT => true
+	];
+
+	private const ALLOWED_BLOCK_IDS = [
+		BlockTypeIds::IRON => true,
+		BlockTypeIds::GOLD => true,
+		BlockTypeIds::DIAMOND => true,
+		BlockTypeIds::EMERALD => true,
+		BlockTypeIds::NETHERITE => true
+	];
+
+	private ?Effect $primaryEffect = null;
+	private ?Effect $secondaryEffect = null;
+
+	public function readStateFromWorld() : Block{
+		parent::readStateFromWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			$this->primaryEffect = EffectIdMap::getInstance()->fromId($tile->getPrimaryEffect());
+			$this->secondaryEffect = EffectIdMap::getInstance()->fromId($tile->getSecondaryEffect());
+		}
+
+		return $this;
+	}
+
+	public function writeStateToWorld() : void{
+		parent::writeStateToWorld();
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if($tile instanceof TileBeacon){
+			if($this->primaryEffect instanceof Effect){
+				$tile->setPrimaryEffect(EffectIdMap::getInstance()->toId($this->primaryEffect));
+			}
+			if($this->secondaryEffect instanceof Effect){
+				$tile->setSecondaryEffect(EffectIdMap::getInstance()->toId($this->secondaryEffect));
+			}
+		}
+	}
+
+	public function getBeaconLevel() : int {
+		$beaconLevel = 0;
+		for($i = self::MIN_LEVEL_BEACON; $i <= self::MAX_LEVEL_BEACON; $i++){
+			if(!$this->isBeaconLevelValid($i)){
+				break;
+			}
+			$beaconLevel++;
+		}
+		return $beaconLevel;
+	}
+
+	public function getPrimaryEffect() : ?Effect{
+		return $this->primaryEffect;
+	}
+
+	/** @return $this */
+	public function setPrimaryEffect(?Effect $primaryEffect) : self{
+		$this->primaryEffect = $primaryEffect;
+		return $this;
+	}
+
+	public function getSecondaryEffect() : ?Effect{
+		return $this->secondaryEffect;
+	}
+
+	/** @return $this */
+	public function setSecondaryEffect(?Effect $secondaryEffect) : self{
+		$this->secondaryEffect = $secondaryEffect;
+		return $this;
+	}
+
 	public function getLightLevel() : int{
 		return 15;
 	}
 
-	//TODO
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($player instanceof Player){
+			$player->setCurrentWindow(new BeaconInventory($this->position));
+		}
+
+		return true;
+	}
+
+	public function onScheduledUpdate() : void{
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 20 * 3);
+
+		if($this->primaryEffect === null){
+			return;
+		}
+		if(($beaconLevel = $this->getBeaconLevel()) >= self::MIN_LEVEL_BEACON){
+			if(!$this->viewSky()){
+				return;
+			}
+
```

Comment:
We need some validation here, it is possible that players might destroy blocks on any level and some effects will be no longer avaible